### PR TITLE
prepare for PR

### PR DESCRIPTION
**Pull request by GitGVP**
This is the initial PR for the realtime_pollen_calibration package written in Python.
The following four files should be reviewed before merging with main is approved:

src/realtime_pollen_calibration/utils.py
src/realtime_pollen_calibration/update_phenology_realtime.py
src/realtime_pollen_calibration/update_strength_realtime.py
README.md
The changes between the two branches have been manually reduced to these four files, as the rest is mostly boilerplate code from the MCH blueprint.

How to code review: https://rewind.com/blog/best-practices-for-reviewing-pull-requests-in-github/
As Guillaume leaves soon the reviewers should consider writing [code suggestions](https://til.hashrocket.com/posts/szgdi4at59-github-insert-a-suggestion-) in addition to their comments.

### CODE REVIEW

File: README.md
Comment (sadamov) [inline README.md:6]: Link to Preprint: https://assets.researchsquare.com/files/rs-2254663/v1/f7f26422c6b4615c2f7e1dfe.pdf?c=1668761631

File: src/realtime_pollen_calibration/update_strength_realtime.py
Comment (GitGVP) [reply src/realtime_pollen_calibration/update_strength_realtime.py:11]: For now it is necessarily one hour. The method of computing the "change" in a field is independant of timestep chosen in its computation but not on its effect (i.e. running the script 6 times an hour, once at each 10 minutes does not produce the same result as running it once).

Allowing to change this timestep is not difficult (i.e. change line 507 in utils.py) but would potentially produce wrong results.

File: src/realtime_pollen_calibration/utils.py
Comment (sadamov) [inline src/realtime_pollen_calibration/utils.py:507]: is this +100 an arbitrary number?

File: src/realtime_pollen_calibration/utils.py
Comment (sadamov) [inline src/realtime_pollen_calibration/utils.py:535]: As mentioned before, I don't think this will work when there are multiple species modelled at the same time, or will it? We will have to make sure that this runs robustly for 1-3 species modelled simultaneously. Either calling this python routine once for all three species or just calling it three times, once for each species seperately.

File: src/realtime_pollen_calibration/utils.py
Comment (sadamov) [inline src/realtime_pollen_calibration/utils.py:552]: Is this function used anywhere? Is it not required because the information comes from the atab file of the measurements?

**Review by sadamov — COMMENTED** [review]
Thanks Guillaume for porting the realtime module to Python. I really appreciated the modular and well-documented codebase, well done! I think the code is much easier to read and understand than my Fortran subroutine tbh. It will also be very smooth to extend the interpolation functionalities in the future and for other met offices to use the code in their own environment. 

In addition to this review we have also carried out a large variety of test cases, even rerunning COSMO and have identified bugs that lead to previous differences between the Python code and the Fortran code. I am confident now that this subroutine can be used operationally after this PR is merged and some software engineering by osm is carried out.

You can find all my comments directly in the code, please accept/deny as you seem fit.

Some general remarks about next steps - @GitGVP these changes are **not** required for the PR to be merged. We will have to find a software engineer at MCH who takes over :)
- We should add a CLI to start this Python library in a shell-script
- Update to the latest copier version and fix the requirements setup
- The boilerplate code from the blueprint that is not needed should be removed (e.g. calculator)
- Tests should be written for all 4 species and all test-cases before going operational
- Additional unit tests are required to assert input and output format of the main functions in this library
- Error handling of the main functions has to be improved - currently the library would crash if input files are missing or something goes wrong during code execution.
- We need to think about versioning and release cycle of this software + define the maintainer after @GitGVP will have left the project

File: src/realtime_pollen_calibration/utils.py
Comment (GitGVP) [reply src/realtime_pollen_calibration/utils.py:507]: Corresponds to +1h for `np.dataTime`

File: src/realtime_pollen_calibration/utils.py
Comment (GitGVP) [reply src/realtime_pollen_calibration/utils.py]: If the station coordinates happen to exactly be on a grid point, you will divide by zero which will cause incorrect results

File: README.md
Comment (sadamov) [reply README.md:39]: We should take this up with the Blueprint team, they should document how to set up mamba if the repo requires it.

**Review by sadamov — APPROVED** [review]
Thanks for the implementation of Named Tuples and handling of Multiple Species. I approve this PR as a first scientific release of the realtime_pollen_calibration library in Python. Cheers :rocket:

File: src/realtime_pollen_calibration/utils.py
Comment (andreaspauling) [reply src/realtime_pollen_calibration/utils.py:93]: As long as the retrieval command does not change, I think yes...
However, the ATAB standard does not guarantee this as there can be optional meta data in the header.

File: src/realtime_pollen_calibration/utils.py
Comment (andreaspauling) [reply src/realtime_pollen_calibration/utils.py:105]: A more robust way could be to add one to the line number of "PARAMETER"

File: src/realtime_pollen_calibration/utils.py
Comment (andreaspauling) [reply src/realtime_pollen_calibration/utils.py:129]: How many times (and more importantly) in how many places does this occur? Still, it is a nice-to-have.

File: src/realtime_pollen_calibration/utils.py
Comment (andreaspauling) [reply src/realtime_pollen_calibration/utils.py:167]: Lets discuss this

File: src/realtime_pollen_calibration/utils.py
Comment (andreaspauling) [reply src/realtime_pollen_calibration/utils.py:341]: There are two unknowns: how often this module is called (now 1h) and how much time it should take for a full adaptation. 24 hours is also kind of arbitrary...
This is work for later

File: src/realtime_pollen_calibration/utils.py
Comment (andreaspauling) [reply src/realtime_pollen_calibration/utils.py:373]: agree. However, values that occur just once are a minor problem. A problem is if the same value has to be changed in 10 different places. This would be error-prone

File: src/realtime_pollen_calibration/utils.py
Comment (GitGVP) [reply src/realtime_pollen_calibration/utils.py]: Done! thank you for your suggeston!

File: src/realtime_pollen_calibration/utils.py
Comment (GitGVP) [reply src/realtime_pollen_calibration/utils.py]: I just moved them at the top of the utils.py file, out of the subroutines. Maybe you will want to move them to a kind of config file and simply load them in the utils.py file but at least now they are only defined once and at the top of a file so they are easy to modify.

File: README.md
Comment (GitGVP) [reply README.md:39]: Agreed with Simon so I resolve this here for now.

**Review by andreaspauling — APPROVED** [review]

**Pull request merged**
